File: test/test_domain.cpp
#include <unity.h>
#include "mocks/MockHardware.h"
#include "domain/Blind.h"

using namespace Blinds::Domain;

MockMotor* motor;
MockLimitSwitch* topLimit;
MockLimitSwitch* bottomLimit;
MockStorage* storage;
Blind* blind;

void setUp(void) {
    motor = new MockMotor();
    topLimit = new MockLimitSwitch();
    bottomLimit = new MockLimitSwitch();
    storage = new MockStorage();

    BlindConfig config;
    config.maxTravelSteps = 1000;
    config.hasLimitSwitches = true;

    blind = new Blind(*motor, topLimit, bottomLimit, *storage, config);
}

void tearDown(void) {
    delete blind;
    delete motor;
    delete topLimit;
    delete bottomLimit;
    delete storage;
}

void test_initialization_unknown(void) {
    storage->savedPos = -1; // Invalid
    blind->initialize();
    TEST_ASSERT_EQUAL(BlindState::UNKNOWN, blind->getState());
}

void test_initialization_restores_pos(void) {
    storage->savedPos = 500;
    blind->initialize();
    TEST_ASSERT_EQUAL(BlindState::IDLE, blind->getState());
    TEST_ASSERT_EQUAL(500, motor->getCurrentPosition());
}

void test_move_up(void) {
    storage->savedPos = 500;
    blind->initialize();

    blind->moveUp();
    TEST_ASSERT_EQUAL(BlindState::MOVING_UP, blind->getState());
    TEST_ASSERT_EQUAL(0, motor->targetPos);
    TEST_ASSERT_TRUE(motor->running);
}

void test_move_down(void) {
    storage->savedPos = 500;
    blind->initialize();

    blind->moveDown();
    TEST_ASSERT_EQUAL(BlindState::MOVING_DOWN, blind->getState());
    TEST_ASSERT_EQUAL(1000, motor->targetPos);
    TEST_ASSERT_TRUE(motor->running);
}

void test_limit_switch_stops_up(void) {
    storage->savedPos = 500;
    blind->initialize();
    blind->moveUp();

    // Simulate hitting limit
    topLimit->triggered = true;
    blind->update();

    TEST_ASSERT_FALSE(motor->running);
    TEST_ASSERT_EQUAL(0, motor->getCurrentPosition());
    TEST_ASSERT_EQUAL(BlindState::IDLE, blind->getState());
}

void test_homing_sequence(void) {
    storage->savedPos = -1;
    blind->initialize();

    blind->startHoming();
    TEST_ASSERT_EQUAL(BlindState::HOMING, blind->getState());
    TEST_ASSERT_TRUE(motor->running);
    // Should be moving down (large positive number)
    TEST_ASSERT_TRUE(motor->targetPos > 1000);

    // Simulate hitting bottom limit
    bottomLimit->triggered = true;
    blind->update();

    // Should stop, set pos to max, and retract immediately
    TEST_ASSERT_TRUE(motor->running);
    TEST_ASSERT_EQUAL(1000, motor->getCurrentPosition());
    TEST_ASSERT_EQUAL(900, motor->targetPos);

    // Simulate retraction finish
    motor->completeMove();
    blind->update();

    TEST_ASSERT_EQUAL(BlindState::IDLE, blind->getState());
}

int main(int argc, char **argv) {
    UNITY_BEGIN();
    RUN_TEST(test_initialization_unknown);
    RUN_TEST(test_initialization_restores_pos);
    RUN_TEST(test_move_up);
    RUN_TEST(test_move_down);
    RUN_TEST(test_limit_switch_stops_up);
    RUN_TEST(test_homing_sequence);
    UNITY_END();
    return 0;
}

File: tests/test_domain.py
import pytest
from src.domain.blind import Blind, BlindConfig, BlindState
from tests.mocks import MockMotor, MockLimitSwitch, MockStorage

@pytest.fixture
def hardware():
    motor = MockMotor()
    top_limit = MockLimitSwitch()
    bottom_limit = MockLimitSwitch()
    storage = MockStorage()
    return motor, top_limit, bottom_limit, storage

@pytest.fixture
def blind(hardware):
    motor, top, bottom, storage = hardware
    config = BlindConfig(max_travel_steps=1000, has_limit_switches=True)
    return Blind(motor, storage, config, top, bottom)

def test_initialization_unknown(blind, hardware):
    _, _, _, storage = hardware
    storage.saved_pos = -1
    blind.initialize()
    assert blind.get_state() == BlindState.UNKNOWN

def test_initialization_restored(blind, hardware):
    motor, _, _, storage = hardware
    storage.saved_pos = 500
    blind.initialize()
    assert blind.get_state() == BlindState.IDLE
    assert motor.get_current_position() == 500

def test_move_up(blind, hardware):
    motor, _, _, storage = hardware
    storage.saved_pos = 500
    blind.initialize()

    blind.move_up()
    assert blind.get_state() == BlindState.MOVING_UP
    assert motor.target_pos == 0
    assert motor.running == True

def test_limit_switch_stop(blind, hardware):
    motor, top, _, storage = hardware
    storage.saved_pos = 500
    blind.initialize()
    blind.move_up()

    top.triggered = True
    blind.update()

    assert motor.running == False
    assert motor.current_pos == 0
    assert blind.get_state() == BlindState.IDLE

def test_homing(blind, hardware):
    motor, _, bottom, storage = hardware
    storage.saved_pos = -1
    blind.initialize()

    blind.start_homing()
    assert blind.get_state() == BlindState.HOMING
    assert motor.running == True

    # Trigger bottom limit
    bottom.triggered = True
    blind.update()

    # Should stop, set max pos, and retract
    assert motor.running == True # Retracting
    assert motor.current_pos == 1000

    # Finish retract
    motor.complete_move()
    blind.update()

    assert blind.get_state() == BlindState.IDLE
